File: Advent of Code/2022/day 3/solution.ts
import * as fs from "fs";

function day_1() {
    let total: number = 0;

    const input = fs.readFileSync("./input.txt", "utf-8");
    const input_array = input.split("\n");

    for (let index = 0; index < input_array.length; index++) {
        let duplicate_items: string[] = [];

        const element = input_array[index];
        const first_rucksack = element.substring(0, (element.length / 2));
        const second_rucksack = element.substring((element.length / 2), element.length);

        // Check if each letter of `first_rucksack` is in `second_rucksack` and not in `duplicate_items
        for (let letter_index = 0; letter_index < first_rucksack.length; letter_index++) {
            const f = first_rucksack[letter_index];

            if (second_rucksack.includes(f) && !duplicate_items.includes(f)) {
                duplicate_items.push(f);
            }
        }

        for (let index = 0; index < duplicate_items.length; index++) {
            const element = duplicate_items[index];

            if (element == element.toLowerCase()) {
                total += element.charCodeAt(0) - 96
            } else {
                total += element.charCodeAt(0) - 38
            }
        }
    }

    console.log(total);
}


function day_2() {
    let total: number = 0;
    let badge_characters: string[] = [];
    let starting_index = 0;
    let ending_index = 2;

    const input = fs.readFileSync("./input.txt", "utf-8");
    const lines_in_file = input.split("\n");

    while (ending_index <= lines_in_file.length) {
        let first_line = lines_in_file[starting_index];
        let second_line = lines_in_file[starting_index + 1];
        let third_line = lines_in_file[ending_index];

        for (let index = 0; index < first_line.length; index++) {
            const element = first_line[index];

            if ((second_line.includes(element)) && (third_line.includes(element))) {
                badge_characters.push(element);
                break;
            }
        }
        starting_index += 3;
        ending_index += 3;
    }

    for (let index = 0; index < badge_characters.length; index++) {
        const element = badge_characters[index];

        if (element == element.toLowerCase()) {
            total += element.charCodeAt(0) - 96
        } else {
            total += element.charCodeAt(0) - 38
        }
    }

    console.log(total);
}


day_1();
day_2();

File: Advent of Code/2022/day 3/solution.py
def part_1():
    total = 0

    with open("./input.txt", "r") as f:
        for line in f:
            duplicate_items = []

            line = line.rstrip()
            first_rucksack = line[0:len(line)//2]
            second_rucksack = line[len(line)//2:]

            for f in first_rucksack:
                if f in second_rucksack and f not in duplicate_items:
                    duplicate_items.append(f)

            for letter in set(duplicate_items):
                if letter.islower():
                    total += (ord(letter) - 96)
                elif letter.isupper():
                    total += ord(letter) - 38

    print(total)


def part_2():
    total = 0
    badge_characters = []
    lines_in_file: list[str] = []
    starting_index = 0
    ending_index = 2

    with open("./input.txt", "r") as f:
        for line in f:
            lines_in_file.append(line.rstrip())

    while ending_index <= len(lines_in_file):
        first_line = lines_in_file[starting_index]
        second_line = lines_in_file[starting_index + 1]
        third_line = lines_in_file[ending_index]

        for character in first_line:
            if ( (character in second_line) and (character in third_line) ):
                badge_characters.append(character)
                break
            else:
                pass
        starting_index += 3
        ending_index += 3

    for letter in badge_characters:
        if letter.islower():
            total += (ord(letter) - 96)
        elif letter.isupper():
            total += ord(letter) - 38

    print(total)


part_1()
part_2()
